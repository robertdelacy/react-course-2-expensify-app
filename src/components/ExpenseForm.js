import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense? props.expense.description : '',
            note: props.expense? props.expense.note : '',
            amount:props.expense? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(props.time),
            calenderFocused: false,
            error:''
        };

        //console.log(this.state)
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState((prevState) => {
            return ({...prevState, description });
        });
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        if(note === this.state.note+'\n'){
            console.log('hit');
            e.stopPropagation();
        }
        this.setState(() => {
            return ({ note });
        });
    };

    onDateChanged = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onAmountChange = (e) => {
        const value = e.target.value;
        if(value.split('£').length > 1){
            const amount = value.split('£')[1];
            if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
                this.setState(() => ({ amount: amount }));
            }
        } else if (!value || !isNaN(parseFloat(value))) {
            this.setState(() => ({ amount: value }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({calenderFocused: focused}))
    };

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }));
        }
        else{
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return(
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.onDescriptionChange}/>
                    <input type="text" placeholder="£" value={ this.state.amount && `£${this.state.amount}`} onChange={this.onAmountChange}/>
                    <SingleDatePicker date={this.state.createdAt}
                    onDateChange={this.onDateChanged}
                    focused={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    displayFormat = 'DD/MM/YYYY'/>
                    <textarea placeholder="Add a note for your expense(optional)" value = {this.state.note} onChange={this.onNoteChange}/>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}