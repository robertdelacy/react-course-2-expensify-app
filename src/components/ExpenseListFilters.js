import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortBy, setStartDate, setEndDate } from '../actions/filters';
import 'react-dates/initialize';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        this.props.sortBy(e.target.value);
    };

    render() {
        return (
            <div>
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
        <select
        value={this.props.filters.sortBy}
        onChange={this.onSortChange}>
            <option value = "description_low">Description (A-Z)</option>
            <option value = "description_high">Description (Z-A)</option>
            <option value = "amount_low">Amount (Small to Large)</option>
            <option value = "amount_high">Amount (Large to Small)</option>
            <option value = "date_low">Date (Old to New)</option>
            <option value = "date_high">Date (New to Old)</option>
        </select>
        <DateRangePicker
        startDate={this.props.filters.startDate}
        startDateId=""
        endDate={this.props.filters.endDate}
        endDateId=""
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1}
        showClearDates={true}
        isOutsideRange={() => false}
        displayFormat = 'DD/MM/YYYY'/>
    </div>
        )
    }
};

const mapStateToProps = (state) => ({
        filters: state.filters
    });

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortBy: (sortByValue) => dispatch(sortBy(sortByValue)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);