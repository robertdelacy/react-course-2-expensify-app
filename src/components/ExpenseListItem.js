import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{amount} - {moment(createdAt).format('Do MMMM, YYYY')}</p>
    </div>
);

export const Test = () => (
    <div>
        <Link to={`/edit/10`}>
            <h3>Coffee</h3>
        </Link>
        <p>Â£3.10 12th February, 2019</p>
    </div>
);

export default ExpenseListItem;