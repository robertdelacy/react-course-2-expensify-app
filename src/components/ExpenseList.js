import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {props.expenses.length === 0 ? ( <h1>No Expenses</h1> ) :
        (<h1>Expense List</h1>,
        props.expenses.map((expense) => <ExpenseListItem key={expense.id}  {...expense} />))}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);