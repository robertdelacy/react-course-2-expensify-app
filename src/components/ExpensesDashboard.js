import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
//import 'react-dates/lib/css/_datepicker.css';

const ExpenseDashboardPage = (props) => (
    <div>
        <ExpenseListFilters/>
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;