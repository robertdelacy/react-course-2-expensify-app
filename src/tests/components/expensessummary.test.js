import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import numeral from 'numeral';

numeral.register('locale', 'gb', {
    delimiters: {
        thousands: ',',
        decimal: '.'
    },
    abbreviations: {
        thousand: 'k',
        million: 'm',
        billion: 'b',
        trillion: 't'
    },
    ordinal : function (number) {
        return number === 1 ? 'st' : 'rd';
    },
    currency: {
        symbol: 'Â£'
    }
});

numeral.locale('gb');

test('should correctly render ExpensesSummary with no expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={0}  expensesTotal={0}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1}  expensesTotal={256}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={23}  expensesTotal={3456789}/>);
    expect(wrapper).toMatchSnapshot();
});