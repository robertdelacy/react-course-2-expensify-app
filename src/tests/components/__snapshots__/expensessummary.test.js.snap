// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
    , totalling 
    £2.56
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
    , totalling 
    £34,567.89
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with no expenses 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
    , totalling 
    £0.00
  </h1>
</div>
`;
